feat: improve waiting for notifications by providing a timeout option

dmigowski opened the pull request:
Improve waiting for notifications using the LISTEN/NOTIFY framework by adding an optional timeout to the PGConnection.getNotifications() method. The change does not have any effects for users that don't use this functionality and the API remain fully backwards compatible.

PS: Please squash commits when merging, the text of this pull request should give a nice commit message.

vlsi commented on pgjdbc/src/main/java/org/postgresql/PGConnection.java:
```diff
@@ -32,6 +32,20 @@
 
   /**
+   * This method returns any notifications that have been received since the last call to this
+   * method. Returns null if there have been no notifications. A timeout can be speficied so the
```
speficied -> specified

dmigowski commented:
OK, I can do that.

Von: Vladimir Sitnikov [mailto:notifications@github.com]
Gesendet: Samstag, 11. März 2017 13:39
An: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Daniel Migowski <dmigowski@ikoffice.de>; Author <author@noreply.github.com>
Betreff: Re: [pgjdbc/pgjdbc] feat: improve waiting for notifications by providing a timeout option (#778)


@vlsi requested changes on this pull request.

Would you please add test cases to cover the new feature?

________________________________

In pgjdbc/src/main/java/org/postgresql/PGConnection.java<https://github.com/pgjdbc/pgjdbc/pull/778#discussion_r105532507>:

> @@ -32,6 +32,20 @@

   PGNotification[] getNotifications() throws SQLException;



   /**

+   * This method returns any notifications that have been received since the last call to this

+   * method. Returns null if there have been no notifications. A timeout can be speficied so the

speficied -> specified

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/778#pullrequestreview-26421764>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADpQwwaCpSyvAMH3mZHhFDCweMJPleHxks5rkpXXgaJpZM4MaKZD>.

davecramer commented:
@dmigowski any chance you can roll this https://github.com/pgjdbc/pgjdbc/pull/579 in at the same time ?

dmigowski commented:
At first I thought about to implement something like NotificationListeners also, but to implement this in a useful way there should be a Thread listening on the Socket all the time, so notifications come along truly asynchronous at the time they are sent. This thread would add incoming messages (except notifications and maybe warnings) to a queue where other Threads read them from, e.g. the QueryExecutor running statements could read its responses from that queue then.

The implementation in #579 is not very useful in my eyes because you still have to issue commands to the server or at least call getNotifications() from time to time to have the notifications delivered. If timely delivery is not a concern you can then dispatch them to your listeners manually, so nothing gained. Does anyone see a nice usecase for these notifications?

Because my first approach with having a distinct reader thread seemed to heavy, I came up with the solution of a timeouting getNotifications(), but this is of course only useful when you use your connection for nothing else but listening for notifies and reacting to them.

More thoughts: If one uses my getNotifications() with timeout on a distinct connection to the database handled by a distinct thread, one could easily write client code that dispatches these notifications.

Conclusion: When you still find the “async notification feature” useful, I would add them here, but as I said, they are not even async in the current implementation. The feature simply describes the dispatching of them on a listener basis.


Von: Dave Cramer [mailto:notifications@github.com]
Gesendet: Samstag, 11. März 2017 22:04
An: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Daniel Migowski <dmigowski@ikoffice.de>; Mention <mention@noreply.github.com>
Betreff: Re: [pgjdbc/pgjdbc] feat: improve waiting for notifications by providing a timeout option (#778)


@dmigowski<https://github.com/dmigowski> any chance you can roll this #579<https://github.com/pgjdbc/pgjdbc/pull/579> in at the same time ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/778#issuecomment-285899971>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADpQwwDDy8fK5m-7cN9KIcAeSKqrfzRKks5rkwxRgaJpZM4MaKZD>.

dmigowski commented:
@vlsi: I added the unit tests, also fixing a bug I found in the mean time. The code coverage is sadly -0.01% now, because i added some code I cannot get to run in the tests (the case of a Warning sent by the server while waiting for notifications). I would fix that if you can tell me how to simulate that in my tests.

dmigowski commented:
Also would like to have my patch committed now. I could create a separate one for #579 if that is still wished for.

davecramer commented:
looks good to me

dmigowski commented:
@vlsi: What is your opinion about this commit now that I added unit tests? What about my discussion regarding your #579?

dmigowski commented:
@vlsi: Can I do something for this change to be accepted? I added unit tests as requested. Please don't be insulted because I did't want to incorporate your #579 this time. Thanks in advance.

davecramer commented:
@dmigowski #579 is mine, he's not insulted by it ;) I'll try to get to this Monday. Thanks!

dmigowski commented:
Thank you very much! If this gets pulled in, I will prepare my next pull request which increases the drivers performance by a few percent points.

Von: Dave Cramer [mailto:notifications@github.com]
Gesendet: Sonntag, 19. März 2017 12:12
An: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Daniel Migowski <dmigowski@ikoffice.de>; Mention <mention@noreply.github.com>
Betreff: Re: [pgjdbc/pgjdbc] feat: improve waiting for notifications by providing a timeout option (#778)


@dmigowski<https://github.com/dmigowski> #579<https://github.com/pgjdbc/pgjdbc/pull/579> is mine, he's not insulted by it ;) I'll try to get to this Monday. Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/778#issuecomment-287609285>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADpQw8GMRThoDkPF9GJkR4tcz0ojxasqks5rnQ1kgaJpZM4MaKZD>.

vlsi commented:
@dmigowski , there is a couple of things that bothers me
1) `setSocketTimeout(0)`. It looks like it would interfere with existing `socketTimeout` connection option.

2) Can you clarify why did you put all the tests into a single test method? Having separate test methods helps debugging (it is easier to execute/debug a test method rather than a part of a big test method)

dmigowski commented:
Hello Vladimir,

I didn’t know that socket timeout was already used on other places, but you are right, I should read the current value and reset it, just in case it has been changed by other means than the options passed as connection parameters.

I will also split the different tests into separate methods.

Regards,
Daniel Migowski

Von: Vladimir Sitnikov [mailto:notifications@github.com]
Gesendet: Montag, 20. März 2017 07:53
An: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Daniel Migowski <dmigowski@ikoffice.de>; Mention <mention@noreply.github.com>
Betreff: Re: [pgjdbc/pgjdbc] feat: improve waiting for notifications by providing a timeout option (#778)


@dmigowski<https://github.com/dmigowski> , there is a couple of things that bothers me

1.      setSocketTimeout(0). It looks like it would interfere with existing socketTimeout connection option.

2.      Can you clarify why did you put all the tests into a single test method? Having separate test methods helps debugging (it is easier to execute/debug a test method rather than a part of a big test method)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/778#issuecomment-287688069>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADpQw6Vq2tEbjVNNBJlUJE2WNx9Hc46bks5rniJigaJpZM4MaKZD>.

dmigowski commented:
@vlsi: I broke up the test into pieces now and revert the timeout to the previous value after returing from getNotifications().

dmigowski commented:
There is a failure in the replication test suite now that I didn't touch. Could someone please investigate? I don't think it is my fault.

Tests run: 56, Failures: 1, Errors: 0, Skipped: 3, Time elapsed: 7.57 sec <<< FAILURE! - in org.postgresql.replication.ReplicationTestSuite

testLastReceiveLSNCorrectOnView(org.postgresql.replication.LogicalReplicationStatusTest)  Time elapsed: 0.063 sec  <<< FAILURE!

java.lang.AssertionError: Replication stream by execute forceUpdateStatus should send to view actual received position that allow monitoring lag

Expected: <LSN{0/16D3390}>
     but: was <LSN{0/16D3457}>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.postgresql.replication.LogicalReplicationStatusTest.testLastReceiveLSNCorrectOnView(LogicalReplicationStatusTest.java:162)

dmigowski commented:
@vlsi: Is there still something to do? @davecramer: Can you pull this in now?

nikiask commented:
Should #779 be closed?
